apps/api/src/app/shared/framework/idempotency.interceptor.ts

```diff
@@ -0,0 +1,242 @@
+  }
+
+  async setCache(
+    key: string,
+    val: { status: ReqStatusEnum; bodyHash: string; data?: any; statusCode?: number },
+    ttl: number,
+    ifNotExists?: boolean
+  ): Promise<string | null> {
+    try {
+      if (ifNotExists) {
+        return await this.cacheService.setIfNotExist(key, JSON.stringify(val), { ttl });
+      }
+      await this.cacheService.set(key, JSON.stringify(val), { ttl });
+    } catch (err) {
+      Logger.warn(`An error occurred while setting idempotency cache, key:${key} error: ${err.message}`, LOG_CONTEXT);
+    }
+
+    return null;
+  }
+
+  private buildError(error: any): HttpException {
+    const statusCode = error.status || error.response?.statusCode || 500;
+    if (statusCode == 500 && !error.response) {
+      //some unhandled exception occured
+      return new InternalServerErrorException();
+    }
+
+    return new HttpException(error.response || error.message, statusCode, error.response?.options);
+  }
+
+  private setHeaders(response: any, headers: Record<string, string>) {
+    Object.keys(headers).map((key) => {
+      if (headers[key]) {
+        response.set(key, headers[key]);
+      }
+    });
+  }
+
+  private hashRequestBody(body: object): string {
+    const hash = createHash('blake2s256');
+    hash.update(Buffer.from(JSON.stringify(body)));
+
+    return hash.digest('hex');
+  }
+
+  private async handlerDuplicateRequest(context: ExecutionContext, bodyHash: string): Promise<Observable<any>> {
+    const cacheKey = this.getCacheKey(context);
+    const idempotencyKey = this.getIdempotencyKey(context)!;
+    const data = await this.cacheService.get(cacheKey);
+    this.setHeaders(context.switchToHttp().getResponse(), { [HEADER_KEYS.IDEMPOTENCY_KEY]: idempotencyKey });
+    const parsed = JSON.parse(data);
+    if (parsed.status === ReqStatusEnum.PROGRESS) {
+      // api call is in progress, so client need to handle this case
+      Logger.error(`previous api call in progress rejecting the request. key:${idempotencyKey}`, LOG_CONTEXT);
+      const error = IdempotencyConflictEnum.REQUEST_PROCESSING;
+      this.setHeaders(context.switchToHttp().getResponse(), {
+        [HEADER_KEYS.IDEMPOTENCY_CONFLICT]: error,
+        [HEADER_KEYS.RETRY_AFTER]: `1`,
+      });
+
+      return throwError(
+        () =>
+          new HttpException(
+            `Request with key "${idempotencyKey}" is currently being processed. Please retry after 1 second`,
+            429
+          )
+      );
+    }
+    if (bodyHash !== parsed.bodyHash) {
+      //different body sent than before
+      Logger.error(`idempotency key is being reused for different bodies. key:${idempotencyKey}`, LOG_CONTEXT);
+      this.setHeaders(context.switchToHttp().getResponse(), {
+        [HEADER_KEYS.IDEMPOTENCY_CONFLICT]: IdempotencyConflictEnum.DIFFERENT_BODY,
+      });
+
+      return throwError(
+        () => new ConflictException(`Request with key "${idempotencyKey}" is being reused for a different body`)
```

Comment:
In alignment with [Idempotency-Key header spec](https://datatracker.ietf.org/doc/html/draft-ietf-httpapi-idempotency-key-header-00#section-2.7) and in moving toward the specification, I'd suggest we return the recommended `422 Unprocessable entity`. If we move ahead with this approach, we can also remove the custom `x-idempotency-conflict` which is non-compliant, the client is instead informed of error scenarios via presence of either `409 Conflict` or `422 Unprocessable entity` as described in the afformentioned spec.

@Cliftonz in dealing with business-critical cases that may be impacted by the small likelihood of incidental `Idempotency-Key` conflict on different bodies, I'd suggest we update our docs to recommend generation of extremely high entropy keys, such as a double UUID `${uuidv4()}${uuidv4()}`. That way we keep our implementation minimal and force the client to deal with entropy, because this really is a client-side error and high-entropy generation tools will likely get better with time. Keen to hear your thoughts too @mahendraHegde.

Reply:
Yes, makes sense, its client side concern, we could reduce the collision by including more parameters in the cache key such as resource path and http method etc, but at the end of the day its client's responsibility to generate collision free keys

Reply:
i have made rest of the changes except removal of `x-idempotency-conflict` , let me know if this is needed as well and if you have other thoughts on the PR.

Reply:
Let's also remove the `x-idempotency-conflict` header as it will be a breaking change if we need to remove it later.